operators/src/upgrade/rest/service.rs

```diff
@@ -125,9 +138,179 @@ pub async fn apply_upgrade() -> Result<HttpResponse, RestError> {
+        let node_name = pod
+            .spec
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("pod.spec.node_name is invalid"))?
+            .node_name
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("pod.spec.node_name is invalid"))?
+            .as_str();
+
+        tracing::info!(
+            "Upgrade starting for data-plane pod: {:#?} present on node: {:#?} ",
+            pod.metadata.name,
+            node_name
+        );
+
+        // Drain the node
+        match UpgradeConfig::get_config()
+            .rest_client()
+            .nodes_api()
+            .put_node_drain(node_name, DRAIN_FOR_UPGRADE)
+            .await
+        {
+            Ok(_) => {
+                // Wait for any node drain process to complete across the cluster.
+                while is_draining().await? {
+                    tokio::time::sleep(Duration::from_secs(10_u64)).await;
+                }
+                tracing::info!("Mayastor node: {:#?} drained successfully", node_name);
+
+                // Wait for 30 seconds for any rebuilds to kick in.
+                tokio::time::sleep(Duration::from_secs(30_u64)).await;
+
+                // wait for the rebuild to cpmplete if any
+                while is_rebuilding().await? {
+                    tokio::time::sleep(Duration::from_secs(10_u64)).await;
+                }
+
+                // Deleting the io-engine pod
+                let pod_name = pod.metadata.name.as_ref().unwrap();
+                tracing::info!(
+                    "Deleting io-engine pod: {:#?} present on node: {:#?}",
+                    pod_name,
+                    node_name,
+                );
+
+                pods.delete(pod_name, &DeleteParams::default())
+                    .await?
+                    .map_left(|o| tracing::debug!("Deleting Pod: {:?}", o.status))
+                    .map_right(|s| tracing::info!("Deleted Pod: {:?}", s));
+
+                // Uncordon the drained node
+                match UpgradeConfig::get_config()
+                    .rest_client()
+                    .nodes_api()
+                    .delete_node_cordon(node_name, DRAIN_FOR_UPGRADE)
+                    .await
+                {
+                    Ok(_) => {
+                        tracing::info!("Node {:#?} uncordoned!", node_name);
+                    }
+                    Err(error) => {
+                        return Err(error.into());
+                    }
+                }
+                // Validate the new pod is up and running
+                while is_data_plane_pod_running(node_name).await? {
+                    tokio::time::sleep(Duration::from_secs(10_u64)).await;
+                }
+                // Validate the control plane pod is up and running
+                is_control_plane_running().await?
+            }
+            Err(error) => {
+                return Err(error.into());
+            }
+        }
+    }
+    Ok(())
+}
```

Comment:
How about using a state machine here, would making the boundaries of each step clearer?

Reply:
What do you mean by state machine here ?

Reply:
State machine itself is probably not really necessary, simply making each state its own function for example would already help tidy things up here I think, eg states:
Drain Node, WaitDrain, RestartDataPlane, etc...
Then we could have:
```rust
let node_name = "node-1";
issue_node_drain(&node_name).await?;
wait_node_drain(&node_name).await?;
restart_data_plane(&node_name).await?;
```

Reply:
done , thanks